enhancements/cluster-validation-tool.md

```diff
@@ -0,0 +1,521 @@
+   the required CRs in the group are included then no report of "missing content" for the group will
+   be generated. Optionally the tool may allow for explicit include/exclude of a component-name by
+   command line argument.
+1. The `required` field indicates if the CR must be present in the input configuration. Any required
+   CR which is not included is reported in the summary as "missing content". The required field has
+   a scope of "within the component-name". If a specfici component-name is not included in the
+   analysis then any required CRs within that component are not flagged as missing content.
+
+| key | value | default |
+| --- | ---- | --- |
+| part-of | the highest level that the CR is part-of e.g DU. | "" |
+| required | must be available. | true |
+| component-name | bundle up sibling CRs. E.g ref PerformaceProfile and TuneD. | "" |
+
+```yaml
+metadata:
+  annotations:
+    reference-config.openshift.io/attributes: |
+      {"part-of": "sample.reference.config",
+       "required": true,
+       "component-name": "someSubComponent"
+       }
+```
+
+#### Example Reference Configuration CR
+User variable content is handled by golang formatted templating within the reference configuration
+CRs. This templating format allows for simple "any value", complex validation, and conditional
+inclusion/exclusion of content. The following types of user variation are expected to be handled:
+1. Mandatory user-defined fields. Examples are marked #1.
+1. Optional user-defined fields. Examples are marked #2
+1. Validation of user defined fields. Examples are marked with #3-n
+
+```yaml
+# PerformanceProfile.yaml with Validation
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-performance-profile
+  labels:
+    # legacy label. This may be obsoleted by the new annotation below.
+    ran.openshift.io/reference-configuration: "du"
+  annotations:
+    # new annotation. describe in a previous section
+    reference-config.openshift.io/attributes: |
+      {"part-of": "telco.ran.du",
+       "required": true,
+       "component-name": "performance"}
+spec:
+  additionalKernelArgs: # This entire block is an example of required "fixed" content
+    - "rcupdate.rcu_normal_after_boot=0"
+    - "efi=runtime"
+    - "module_blacklist=irdma"
+  cpu:
+    isolated: {{ .spec.cpu.isolated  }} # #1 mandatory user variable content
+    reserved: {{ .spec.cpu.reserved  }} # #1 mandatory user variable content
+  hugepages:
+    defaultHugepagesSize: {{ .spec.hugepages.defaultHugepagesSize  }} # #1 mandatory
+    pages:
+      {{- maxListLength 1   .spec.hugepages.pages }} #3-1 defined in Go
+        {{- range .spec.hugepages.pages }}
+      - size:  {{block "validatesize" .}} {{ .size  }} {{end}} #3-2 can be defined later during runtime or falls back to .size
+        count: {{ .count }} {{ if eq "float64" (printf "%T" .count) }}is a float64 {{ else }} is not a float64{{ end }}
```

Comment:
This is coding already and should never be exposed like this.

Reply:
I agree. This is not a good example of what we expect in the reference. It was included to show what _can_ be done if/when required but this level of "coding" is not the mainline use or expectation. I'll update the enhancement to remove this and add a better example here.